### PR TITLE
[IOTDB-1762] IoTDB REST Data Services : OpenAPI Support

### PR DESCRIPTION
**Pull request by CloudWise-Lukemiao**
iotdb add  OpenAPI module

### CODE REVIEW

File: openapi/java-rest/pom.xml [hunk 1]
```diff
@@ -0,0 +1,597 @@
...
+                    <groupId>com.brsanthu</groupId>
+                    <artifactId>migbase64</artifactId>
+                    <version>2.2</version>
+                </dependency>
+                <dependency>
+                    <groupId>com.google.protobuf</groupId>
+                    <artifactId>protobuf-java</artifactId>
+                    <version>3.5.1</version>
+                </dependency>
+                <!-- Bean Validation API support -->
```
Comment (qiaojialin) [inline openapi/java-rest/pom.xml:406]: remove unused dependency

File: openapi/java-rest/pom.xml [hunk 2]
```diff
@@ -0,0 +1,597 @@
...
+                                    </sources>
+                                </configuration>
+                            </execution>
+                        </executions>
+                    </plugin>
+                </plugins>
+            </build>
+            <dependencies>
+                <dependency>
+                    <groupId>io.swagger</groupId>
```
Comment (qiaojialin) [inline openapi/java-rest/pom.xml:357]: please check if the license is forbidden by apache.
http://www.apache.org/legal/resolved.html#category-x

File: openapi/java-rest/pom.xml [hunk 3]
```diff
@@ -0,0 +1,597 @@
...
+                    <artifactId>servlet-api</artifactId>
+                    <version>${servlet-api-version}</version>
+                </dependency>
+                <dependency>
+                    <groupId>com.fasterxml.jackson.datatype</groupId>
+                    <artifactId>jackson-datatype-jsr310</artifactId>
+                    <version>${jackson-version}</version>
+                </dependency>
+                <dependency>
+                    <groupId>com.fasterxml.jackson.jaxrs</groupId>
```
Comment (qiaojialin) [inline openapi/java-rest/pom.xml:391]: check license

File: openapi/java-rest/src/main/java/org/apache/iotdb/openapi/gen/handler/factories/V1ApiServiceFactory.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+ */
+
+// this file is generated by openapi generator
+
+package org.apache.iotdb.openapi.gen.handler.factories;
+
+import org.apache.iotdb.openapi.gen.handler.V1ApiService;
+import org.apache.iotdb.openapi.gen.handler.impl.V1ApiServiceImpl;
+
+@javax.annotation.Generated(
```
Comment (qiaojialin) [inline openapi/java-rest/src/main/java/org/apache/iotdb/openapi/gen/handler/factories/V1ApiServiceFactory.java:25]: is this needed?

File: openapi/java-rest/src/main/java/org/apache/iotdb/openapi/gen/handler/impl/ProtobufMessageBodyWriter.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+@Provider
+@Produces("application/x-protobuf")
+public class ProtobufMessageBodyWriter implements MessageBodyWriter<Message> {
+
+  @Override
+  public boolean isWriteable(
+      Class<?> aClass, Type type, Annotation[] annotations, MediaType mediaType) {
+    return Message.class.isAssignableFrom(aClass);
+    // return false;
```
Comment (qiaojialin) [inline openapi/java-rest/src/main/java/org/apache/iotdb/openapi/gen/handler/impl/ProtobufMessageBodyWriter.java:44]: remove

File: openapi/java-rest/src/main/java/org/apache/iotdb/openapi/gen/handler/impl/ProtobufMessageBodyWriter.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    ByteArrayOutputStream baos = new ByteArrayOutputStream();
+    try {
+      message.writeTo(baos);
+    } catch (IOException e) {
+      return -1;
+    }
+    byte[] bytes = baos.toByteArray();
+    buffer.put(message, bytes);
+    return bytes.length;
+    // return 0;
```
Comment (qiaojialin) [inline openapi/java-rest/src/main/java/org/apache/iotdb/openapi/gen/handler/impl/ProtobufMessageBodyWriter.java:61]: remove

File: openapi/java-rest/src/main/java/org/apache/iotdb/openapi/gen/handler/impl/V1ApiServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+  public Response postV1GrafanaData(
+      GroupByFillPlan groupByFillPlan, SecurityContext securityContext) throws NotFoundException {
+    // do some magic!
+    return Response.ok().entity(new ApiResponseMessage(ApiResponseMessage.OK, "magic!")).build();
+  }
+
+  @Override
+  public Response postV1GrafanaDataSimplejson(
+      GroupByFillPlan groupByFillPlan, SecurityContext securityContext) throws NotFoundException {
+    // do some magic!
```
Comment (qiaojialin) [inline openapi/java-rest/src/main/java/org/apache/iotdb/openapi/gen/handler/impl/V1ApiServiceImpl.java:47]: remove

File: server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java [hunk 8]
```diff
@@ -729,6 +729,100 @@
...
+
+  /** set the OpenApi reset port. */
+  private int OpenApiPort = 18080;
+
+  /** set the OpenApi storegroup number. */
+  private int sgCount = 5;
+
+  /** enable the OpenApi ssl. */
+  private boolean enable_https = false;
+  /** openapi ssl key Store Path */
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java:743]: add an empty line

File: server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java [hunk 9]
```diff
@@ -729,6 +729,100 @@
 
   private String adminPassword = "root";
 
+  /** if the startOpenApi is true, we will start OpenApi */
+  private boolean startOpenApi = true;
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java:733]: the default could be false, be consistent with iotdb-engine.properties

File: server/src/main/java/org/apache/iotdb/openapi/gen/handler/impl/V1ApiServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,1025 @@
...
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+
+@javax.annotation.Generated(
+    value = "org.openapitools.codegen.languages.JavaJerseyServerCodegen",
+    date = "2021-01-13T21:45:03.765+08:00[Asia/Shanghai]")
+public class V1ApiServiceImpl extends V1ApiService {
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/openapi/gen/handler/impl/V1ApiServiceImpl.java:89]: why  V1?

**Review by qiaojialin — COMMENTED** [review]
Please add a UserGuide in docs/Userguide

File: server/src/main/java/org/apache/iotdb/openapi/gen/handler/impl/V1ApiServiceImpl.java (hunk 10 above)
Comment (CloudWise-Lukemiao) [reply server/src/main/java/org/apache/iotdb/openapi/gen/handler/impl/V1ApiServiceImpl.java:89]: The method name is the code automatically generated from the configuration file  iotdb-rest.yaml

File: openapi/java-rest/src/main/java/org/apache/iotdb/openapi/gen/handler/impl/ProtobufMessageBodyWriter.java (hunk 5 above)
Comment (CloudWise-Lukemiao) [reply openapi/java-rest/src/main/java/org/apache/iotdb/openapi/gen/handler/impl/ProtobufMessageBodyWriter.java]: fixed

File: openapi/java-rest/src/main/java/org/apache/iotdb/openapi/gen/handler/impl/ProtobufMessageBodyWriter.java (hunk 6 above)
Comment (CloudWise-Lukemiao) [reply openapi/java-rest/src/main/java/org/apache/iotdb/openapi/gen/handler/impl/ProtobufMessageBodyWriter.java]: fixed

File: openapi/java-rest/src/main/java/org/apache/iotdb/openapi/gen/handler/impl/V1ApiServiceImpl.java (hunk 7 above)
Comment (CloudWise-Lukemiao) [reply openapi/java-rest/src/main/java/org/apache/iotdb/openapi/gen/handler/impl/V1ApiServiceImpl.java]: fixed

File: server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java (hunk 8 above)
Comment (CloudWise-Lukemiao) [reply server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java:744]: fixed

File: server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java (hunk 9 above)
Comment (CloudWise-Lukemiao) [reply server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java:733]: fixed

File: openapi/java-rest/pom.xml (hunk 3 above)
Comment (CloudWise-Lukemiao) [reply openapi/java-rest/pom.xml:384]: apache license

File: openapi/java-rest/pom.xml (hunk 2 above)
Comment (CloudWise-Lukemiao) [reply openapi/java-rest/pom.xml:350]: apache license

File: openapi/java-rest/pom.xml (hunk 1 above)
Comment (CloudWise-Lukemiao) [reply openapi/java-rest/pom.xml]: fixed

File: openapi/java-rest/src/main/java/org/apache/iotdb/openapi/gen/handler/factories/V1ApiServiceFactory.java (hunk 4 above)
Comment (CloudWise-Lukemiao) [reply openapi/java-rest/src/main/java/org/apache/iotdb/openapi/gen/handler/factories/V1ApiServiceFactory.java]: fixed

File: docs/zh/UserGuide/Communication-Service-Protocol/OpenApi.md [hunk 11]
```diff
@@ -0,0 +1,442 @@
...
+    software distributed under the License is distributed on an
+    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+    KIND, either express or implied.  See the License for the
+    specific language governing permissions and limitations
+    under the License.
+
+-->
+
+# Rest API
+IoTDB的restApi设计用于支持与Grafana和Prometheus的集成同时也提供了查询、插入和non-query接口,它使用OpenAPI标准来定义接口和生成框架源代码。
```
Comment (qiaojialin) [inline docs/zh/UserGuide/Communication-Service-Protocol/OpenApi.md:23]:
```suggestion
IoTDB 的 restApi 设计用于支持与 Grafana 和 Prometheus 的集成同时也提供了读写等接口,它使用 OpenAPI 标准来定义接口和生成框架源代码。
```

File: docs/zh/UserGuide/Communication-Service-Protocol/OpenApi.md [hunk 12]
```diff
@@ -0,0 +1,442 @@
...
+为了IoTDB数据库安全我们建议使用一台非IoTDB的服务器安装反向代理服务（例如nginx等）把http请求转发到IoTDB，当然你可以不使用反向代理服务直接使用openapi的rest，如果你正在使用grafana服务同时使用nginx则需要在nginx.conf添加 add_header 'Access-Control-Max-Age' xx 来保证可以正常使用
```
Comment (qiaojialin) [inline docs/zh/UserGuide/Communication-Service-Protocol/OpenApi.md:24]:
```suggestion
为了 IoTDB 数据库安全我们建议使用一台非 IoTDB 的服务器安装反向代理服务（例如 nginx 等）把 http 请求转发到 IoTDB，当然你可以不使用反向代理服务直接使用 openapi 的 rest，如果你正在使用 grafana 服务同时使用 nginx 则需要在 nginx.conf 添加 add_header 'Access-Control-Max-Age' xx 来保证可以正常使用
```

File: docs/zh/UserGuide/Communication-Service-Protocol/OpenApi.md [hunk 13]
```diff
@@ -0,0 +1,442 @@
...
+OpenApi 接口使用了基础（basic）鉴权，每次url请求都需要在header中携带 'Authorization': 'Basic ' + base64.encode(username + ':' + password)例如:
```
Comment (qiaojialin) [inline docs/zh/UserGuide/Communication-Service-Protocol/OpenApi.md:25]:
```suggestion
OpenApi 接口使用了基础（basic）鉴权，每次url请求都需要在 header 中携带 'Authorization': 'Basic ' + base64.encode(username + ':' + password)例如:
```

File: server/src/test/java/org/apache/iotdb/openapi/apache/iotdb/OpenApiServiceTest.java [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+    BigDecimal stime = BigDecimal.valueOf(1);
+    BigDecimal etime = BigDecimal.valueOf(10);
+    List<String> path = new ArrayList<String>();
+    path.add("root");
+    path.add("sg");
+    path.add("aa");
+    groupByFillPlan.setStime(stime);
+    groupByFillPlan.setEtime(etime);
+    groupByFillPlan.setPaths(path);
+    // groupByFillPlan.set("1ms");
```
Comment (qiaojialin) [inline server/src/test/java/org/apache/iotdb/openapi/apache/iotdb/OpenApiServiceTest.java:65]: remove

File: docs/zh/UserGuide/Communication-Service-Protocol/OpenApi.md [hunk 15]
```diff
@@ -0,0 +1,442 @@
...
+| Labels  |  array | tag信息 |
+|  Name | String|  tagKey |
+| Value  |  array| tagValue |
+| Samples  |  array |   |
+| Timestamp  |  number|  时间 |
+| Value  |  number | 值 |
+
+## 其他接口
+
+###  read 接口
```
Comment (qiaojialin) [inline docs/zh/UserGuide/Communication-Service-Protocol/OpenApi.md:368]:
```suggestion
###  SQL 查询接口
```

File: docs/zh/UserGuide/Communication-Service-Protocol/OpenApi.md [hunk 16]
```diff
@@ -0,0 +1,442 @@
...
+
+返回参数:
+
+|参数名称  |参数类型  |参数描述|
+| ------------ | ------------ | ------------|
+| values | array |  值 |
+| name  |  string | 测点名称 |
+| type | String| 数据类型 |
+
+###  nonQuery 接口
```
Comment (qiaojialin) [inline docs/zh/UserGuide/Communication-Service-Protocol/OpenApi.md:392]:
```suggestion
###  SQL 非查询接口
```

File: docs/zh/UserGuide/Communication-Service-Protocol/OpenApi.md [hunk 17]
```diff
@@ -0,0 +1,442 @@
...
+
+
+返回参数:
+
+|参数名称  |参数类型  |参数描述|
+| ------------ | ------------ | ------------|
+| code | integer |  状态码 |
+| message  |  string | 信息提示 |
+
+###  write 接口
```
Comment (qiaojialin) [inline docs/zh/UserGuide/Communication-Service-Protocol/OpenApi.md:416]:
```suggestion
###  写入接口
```

File: server/src/test/java/org/apache/iotdb/openapi/apache/iotdb/OpenApiServiceTest.java (hunk 14 above)
Comment (CloudWise-Lukemiao) [reply server/src/test/java/org/apache/iotdb/openapi/apache/iotdb/OpenApiServiceTest.java:65]: fixed

**Pull request closed without merging**